test/Test/Database/LSMTree/Normal/StateMachine.hs

```diff
@@ -295,39 +409,56 @@ instance InLockstep ModelState where
       Duplicate tableVar              -> [SomeGVar tableVar]
 
   arbitraryWithVars ::
-       ModelFindVariables ModelState
-    -> ModelState
-    -> Gen (Any (LockstepAction ModelState))
+       ModelFindVariables (ModelState h)
+    -> ModelState h
+    -> Gen (Any (LockstepAction (ModelState h)))
   arbitraryWithVars = arbitraryActionWithVars (Proxy @(Word64, Word64, Word64))
 
   shrinkWithVars ::
-       ModelFindVariables ModelState
-    -> ModelState
-    -> LockstepAction ModelState a
-    -> [Any (LockstepAction ModelState)]
+       ModelFindVariables (ModelState h)
+    -> ModelState h
+    -> LockstepAction (ModelState h) a
+    -> [Any (LockstepAction (ModelState h))]
   shrinkWithVars = shrinkActionWithVars
 
   tagStep ::
-       (ModelState, ModelState)
-    -> LockstepAction ModelState a
-    -> Val a
+       (ModelState h, ModelState h)
+    -> LockstepAction (ModelState h) a
+    -> Val h a
     -> [String]
   tagStep states action = map show . tagStep' states action
 
-deriving instance Show (Val a)
-deriving instance Eq (Obs a)
-deriving instance Show (Obs a)
+deriving instance Show (SUT.Class.TableConfig h) => Show (Val h a)
+deriving instance Eq (Obs h a)
+deriving instance Show (Obs h a)
+
+{-------------------------------------------------------------------------------
+  Real monad
+-------------------------------------------------------------------------------}
+
+-- | Uses 'WrapSession' such that we can define class instances that mention
+-- 'RealMonad' in the class head. This is necessary because class heads can not
+-- mention type families directly.
+--
+-- Also carries an exception handle that is specific to the table implementation
+-- identified by the table handle @h@.
+type RealMonad h m = ReaderT (WrapSession h m, Handler m (Maybe Model.Err)) m
```

Comment:
INFO: each implementation (identified by @h@) could throw different errors, so a handler has to be passed in up front. The handler could probably also be passed into the statemachine framework using a type class